proposals/0016-simd-application-fees.md

```diff
@@ -0,0 +1,438 @@
+or add any kind of congestion control mechanism.
+
+## Alternatives Considered
+
+* Having a fixed write lock fee and a read lock fee.
+
+Pros: Simpler to implement, Simple to calculate fees.
+
+Cons: Will increase fees for everyone, dapp cannot decide to rebate fees, (or
+not all existing apps have to develop a rebate system). Fees cannot be decided
+by the dapp developer. Penalizes everyone for a few bad actors.
+
+* Passing application fees for each account by instruction and validating them
+  during the execution inside user application code
+
+Pros: High efficiency, minimal performance impact, more generic, no need to
+change the account structure.
+
+Cons: Cannot prevent denial of service attacks, or account blocking
+attacks that do not call into the respective program
+
+## New Terminology
+
+Application Fees: Fees that are decided by the dapp developer will be charged if
+a user wants to use the dapp. They will be applied even if the transaction fails
+contrary to lamport transfers. The program can decide to rebate these fees back
+to the user if certain conditions decided by dapp developers are met.
+
+### Other terminology
+
+`Owner` of an account: It is the account owner as specified in the `owner`
+field in the account structure. In the case of an externally owned account
+(i.e, keypair) owner is a `system program` and in case of a Program derived
+address owner is usually the program.
+
+Account `Authority`: The true authority for the account. Let's take an example
+of associated token account. A associated token account is a `PDA` derived
+from associated token program, where the owner is token program. But the
+program internally saves who has the authority over the account. Only token
+program can change the data of a token account. For operations like
+withdrawing tokens, the authority of the associated token account has to be
+the signer for the transaction containing transfer instruction. To receive
+tokens from another account, the authority's signature is not required.
+
+`Other Fees`: Fees other than application fees
+`base fees + prioritization fees`.
+
+## Detailed Design
+
+Application fees enable dapp developer to decide who to rebate and who to
+penalize through runtime features. In particular, programs needs to be able
+to penalize, even if the application is not CPI'd to.
+
+The checks on the application fees will be taken care of by Solana runtime.
+Total application fees paid will be included in the transaction through a
+special `PayApplicationFees` similar to the `ComputeBudget` program so that it
+is easier to calculate the total amount of fees that will be paid. In case of
+nonpayment or partial payment, the program is never executed. This instruction
+cannot be CPI'ed into.
+
+Application developers will be able to rebate these fees through a new runtime
+mechanism, but only if the transaction actually succeeds, this way CU heavy
+simulation of the program while write-locking the respective accounts will
```

Comment:
I'm a little confused by what "CU heavy simulation" means?

Reply:
you can link the owner program in your own program and avoid sending an instruction that would fail by simulating the instructions code. a before mentioned common case of this would be to parse an orderbook or clamm ranges and decide if to send a taker order. arb bots on eth simulate off-chain in custom evm implementations, arb bots on solana just simulate on-chain. will include the term in the motivation.